astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
+        Unit("1 / m")
+
+    This method also allows that the result of unrelated objects is tried
+    with what this class implements as arithmetics and give a result that
+    is the same class as the class that called the method.
+
+        >>> nd2 = nd.ic_multiplication([1,2],3)
+        >>> nd2.__class__.__name__
+        'NDDataWithMath'
+        >>> nd2.data
+        array([3, 6])
+
+    Since this is a classmethod we don't need an instance to use them:
+
+        >>> nd3 = NDDataWithMath.ic_subtraction(5, NDData([1,2,3]))
+        >>> nd3.__class__.__name__
+        'NDDataWithMath'
+        >>> nd3.data
+        array([4, 3, 2])
+
+    And it allows to handle arithmetics with different subclasses.
+
+        >>> class NDDataWithMathAndSlicing(NDSlicingMixin,NDArithmeticMixin, NDData): pass
+        >>> nd = NDDataWithMath([5,5,5])
+        >>> nd2 = NDDataWithMathAndSlicing([3,2,5])
+        >>> nd3 = nd2.ic_addition(nd, nd2)
+        >>> nd3.__class__.__name__
+        'NDDataWithMathAndSlicing'
+        >>> nd3.data
+        array([ 8,  7, 10])
     """
-    def _arithmetic(self, operand, propagate_uncertainties, name, operation):
+
+    def _arithmetic(self, operation, operand,
+                    uncertainty_correlation=0, meta_kwds_operate=None,
+                    meta_kwds_set=None, propagate_uncertainties=True,
+                    handle_mask=True, handle_meta=True, compare_wcs=True,
+                    **kwds):
         """
-        {name} another dataset (``operand``) to this dataset.
+        Base method which calculates the result of the arithmetic operation.
+
+        This class determines the result of the arithmetic operation on the
+        ``data`` with respect to their units and then forwards to other methods
+        to calculate the other properties for the result (like uncertainty).
+
+        TODO: Simplify some parameter names, descriptions or presence? (embray)
+        TODO: Maybe define the parameter descriptions as strings before so that
+        this parameters description does not need to reference other methods...
+        Or simply call them kwargs and say a full description of the parameters
+        is given in method add?
 
         Parameters
         ----------
-        operand : `~astropy.nddata.NDData`
-            The second operand in the operation a {operator} b
+        operation: str
+            The operation that is performed on the `NDData`. Supported are
+            ``addition``, ``subtraction``, ``multiplication`` and ``division``.
+
+            TODO: Use UFUNC instead of string? (embray, mwcraig)
+            I didn't want to do it since each attribute can interpret the
+            function that is used by the name of the operation better than by
+            the function itself. Suppose we have ``data`` that is a
+            `~numpy.matrix` or ``Pandas.DataFrame`` we need to use different
+            UFUNCS (maybe) and that explodes the number of if/elif checks in
```

Comment:
> `add` should provide the numpy ufunc (`np.add`) and pass it to `_arithmetic` as operation? So that another ufunc could be provided by calling `_arithmetic` directly?

Mostly, yes. I think what I said before about substituting a different function without subclassing doesn't actually work, but a subclass that wanted to use a different addition could do so by just overriding `add` without needing to change anything in `_arithmetic`.

The other advantage of making this change is that you eliminate the need for the duplicate logic in `_arithmetic_data` and `_arithmetic_meta`